Vertx-web-client  setTryUseCompression error:decompression failure

hello,
I use webclient request "http://api.bilibili.com/x/v1/dm/list.so?oid=87113082"

code :
```
val userAgent = "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.106 Safari/537.36"
    val webClientOptions = webClientOptionsOf(userAgent=userAgent,followRedirects = true,tryUseCompression = true)
    val weblient = WebClient.create(vertx,webClientOptions)
val response = webClient
        .get(bilibiliApiUrl, "/x/v1/dm/list.so")
        .putHeader("Accept-Encoding","gzip, deflate")
        .addQueryParam("oid", cid.toString())
        .addQueryParam("page", i.toString())
        .sendAwait()
```
But there was an error：
```

io.netty.handler.codec.compression.DecompressionException: decompression failure
        at io.netty.handler.codec.compression.JdkZlibDecoder.decode(JdkZlibDecoder.java:230)
        at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502)
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965)
        at io.netty.channel.embedded.EmbeddedChannel.writeInbound(EmbeddedChannel.java:325)
        at io.netty.handler.codec.http.HttpContentDecoder.decode(HttpContentDecoder.java:234)
        at io.netty.handler.codec.http.HttpContentDecoder.decodeContent(HttpContentDecoder.java:155)
        at io.netty.handler.codec.http.HttpContentDecoder.decode(HttpContentDecoder.java:147)
        at io.netty.handler.codec.http.HttpContentDecoder.decode(HttpContentDecoder.java:46)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:88)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:438)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297)
        at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:253)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:644)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:579)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:496)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:835)
Caused by: java.util.zip.DataFormatException: incorrect header check
        at java.base/java.util.zip.Inflater.inflateBytesBytes(Native Method)
        at java.base/java.util.zip.Inflater.inflate(Inflater.java:378)
        at io.netty.handler.codec.compression.JdkZlibDecoder.decode(JdkZlibDecoder.java:187)
        ... 38 more

```

--- a/src/main/java/io/vertx/core/http/impl/HttpChannelConnector.java
+++ b/src/main/java/io/vertx/core/http/impl/HttpChannelConnector.java
@@ -191,7 +191,7 @@ class HttpChannelConnector implements ConnectionProvider<HttpClientConnection> {
       false,
       options.getDecoderInitialBufferSize()));
     if (options.isTryUseCompression()) {
-      pipeline.addLast("inflater", new HttpContentDecompressor(true));
+      pipeline.addLast("inflater", new HttpContentDecompressor());
     }
   }
 

--- a/src/main/java/io/vertx/core/http/impl/HttpChannelConnector.java
+++ b/src/main/java/io/vertx/core/http/impl/HttpChannelConnector.java
@@ -191,7 +191,7 @@ class HttpChannelConnector implements ConnectionProvider<HttpClientConnection> {
       false,
       options.getDecoderInitialBufferSize()));
     if (options.isTryUseCompression()) {
-      pipeline.addLast("inflater", new HttpContentDecompressor());
+      pipeline.addLast("inflater", new HttpContentDecompressor(false));
     }
   }
 

--- a/src/main/java/io/vertx/core/http/impl/HttpServerImpl.java
+++ b/src/main/java/io/vertx/core/http/impl/HttpServerImpl.java
@@ -393,7 +393,7 @@ public class HttpServerImpl implements HttpServer, Closeable, MetricsProvider {
     pipeline.addLast("httpDecoder", new VertxHttpRequestDecoder(options));
     pipeline.addLast("httpEncoder", new VertxHttpResponseEncoder());
     if (options.isDecompressionSupported()) {
-      pipeline.addLast("inflater", new HttpContentDecompressor(true));
+      pipeline.addLast("inflater", new HttpContentDecompressor(false));
     }
     if (options.isCompressionSupported()) {
       pipeline.addLast("deflater", new HttpChunkContentCompressor(options.getCompressionLevel()));

--- a/src/test/java/io/vertx/core/http/HttpCompressionTest.java
+++ b/src/test/java/io/vertx/core/http/HttpCompressionTest.java
@@ -17,9 +17,6 @@ import io.vertx.core.buffer.Buffer;
 
 import org.junit.Test;
 
-import java.io.ByteArrayOutputStream;
-import java.io.IOException;
-import java.util.zip.Deflater;
 
 /**
  */
@@ -41,7 +38,7 @@ public class HttpCompressionTest extends HttpTestBase {
       " * You may elect to redistribute this code under either of these licenses.\n" +
       " */";
 
-  private HttpServer serverWithMinCompressionLevel, serverWithMaxCompressionLevel,serverWithClientDeCompression = null;
+  private HttpServer serverWithMinCompressionLevel, serverWithMaxCompressionLevel,serverWithClientDeCompression,serverWithServerDecompression = null;
   private HttpClient clientraw = null;
 
   public void setUp() throws Exception {
@@ -55,7 +52,8 @@ public class HttpCompressionTest extends HttpTestBase {
     // server = vertx.createHttpServer();
     serverWithMinCompressionLevel = vertx.createHttpServer(serverOpts.setPort(DEFAULT_HTTP_PORT - 1).setCompressionLevel(1));
     serverWithMaxCompressionLevel = vertx.createHttpServer(serverOpts.setPort(DEFAULT_HTTP_PORT + 1).setCompressionLevel(9));
-    serverWithClientDeCompression = vertx.createHttpServer(serverOpts.setCompressionSupported(true));
+    serverWithClientDeCompression = vertx.createHttpServer(serverOpts.setPort(DEFAULT_HTTP_PORT + 2).setCompressionSupported(true));
+    serverWithServerDecompression = vertx.createHttpServer(serverOpts.setPort(DEFAULT_HTTP_PORT - 2).setDecompressionSupported(true));
   }
 
   @Test
@@ -182,13 +180,31 @@ public class HttpCompressionTest extends HttpTestBase {
     });
 
     startServer(serverWithClientDeCompression);
-    client.get(DEFAULT_HTTP_PORT+1, DEFAULT_HTTP_HOST, DEFAULT_TEST_URI,
+    client.get(DEFAULT_HTTP_PORT+2, DEFAULT_HTTP_HOST, DEFAULT_TEST_URI,
       onSuccess(resp -> {
         resp.bodyHandler(responseBuffer -> {
            assertEquals(compressData,responseBuffer.toString());
           testComplete();
         });
       })).putHeader(HttpHeaders.ACCEPT_ENCODING, HttpHeaders.DEFLATE_GZIP).end();
+    await();
+  }
+
+  @Test
+  public void testServerDecompression() throws Exception {
+    String data = "Test Server Decompression";
+    serverWithServerDecompression.requestHandler(req -> {
+      assertNotNull(req.headers().get("Accept-Encoding"));
+      req.bodyHandler(buf -> {
+        assertEquals(data,buf.toString());
+        testComplete();
+        req.response().putHeader(HttpHeaders.CONTENT_ENCODING,HttpHeaders.DEFLATE_GZIP).end(buf);
+      });
+    });
+    startServer(serverWithServerDecompression);
+    client.get(DEFAULT_HTTP_PORT-2, DEFAULT_HTTP_HOST, DEFAULT_TEST_URI,res -> {
+
+    }).putHeader(HttpHeaders.ACCEPT_ENCODING, HttpHeaders.DEFLATE_GZIP).end(data);
     await();
   }
 }

--- a/src/test/java/io/vertx/core/http/HttpCompressionTest.java
+++ b/src/test/java/io/vertx/core/http/HttpCompressionTest.java
@@ -17,6 +17,10 @@ import io.vertx.core.buffer.Buffer;
 
 import org.junit.Test;
 
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.zip.Deflater;
+
 /**
  */
 public class HttpCompressionTest extends HttpTestBase {
@@ -37,7 +41,7 @@ public class HttpCompressionTest extends HttpTestBase {
       " * You may elect to redistribute this code under either of these licenses.\n" +
       " */";
 
-  private HttpServer serverWithMinCompressionLevel, serverWithMaxCompressionLevel = null;
+  private HttpServer serverWithMinCompressionLevel, serverWithMaxCompressionLevel,serverWithClientDeCompression = null;
   private HttpClient clientraw = null;
 
   public void setUp() throws Exception {
@@ -51,6 +55,7 @@ public class HttpCompressionTest extends HttpTestBase {
     // server = vertx.createHttpServer();
     serverWithMinCompressionLevel = vertx.createHttpServer(serverOpts.setPort(DEFAULT_HTTP_PORT - 1).setCompressionLevel(1));
     serverWithMaxCompressionLevel = vertx.createHttpServer(serverOpts.setPort(DEFAULT_HTTP_PORT + 1).setCompressionLevel(9));
+    serverWithClientDeCompression = vertx.createHttpServer(serverOpts.setCompressionSupported(false));
   }
 
   @Test
@@ -165,5 +170,44 @@ public class HttpCompressionTest extends HttpTestBase {
           && rawMinCompressionResponseByteCount > rawMaxCompressionResponseByteCount);
       testComplete();
     }
+  }
+
+  @Test
+  public void testClientDeCompression() throws Exception {
+    String compressData = "Test Client DeCompression....";
+    serverWithClientDeCompression.requestHandler(req -> {
+      assertNotNull(req.headers().get("Accept-Encoding"));
+      Deflater compress = new Deflater(Deflater.DEFAULT_COMPRESSION, true);
+      byte[] bytes = compressData.getBytes();
+      compress.setInput(bytes);
+      compress.finish();
+      ByteArrayOutputStream bos = new ByteArrayOutputStream(bytes.length);
+      byte[] buf = new byte[1024];
+      byte[] output;
+      while (!compress.finished()) {
+        int i = compress.deflate(buf);
+        bos.write(buf, 0, i);
+      }
+      output = bos.toByteArray();
+      try {
+        bos.close();
+      } catch (IOException e) {
+        e.printStackTrace();
+      }
+      compress.end();
+      req.response()
+        .putHeader(HttpHeaders.CONTENT_ENCODING, "deflate")
+        .end(Buffer.buffer(output));
+    });
+
+    startServer(serverWithClientDeCompression);
+    client.get(DEFAULT_HTTP_PORT+1, DEFAULT_HTTP_HOST, DEFAULT_TEST_URI,
+      onSuccess(resp -> {
+        resp.bodyHandler(responseBuffer -> {
+           assertEquals(compressData,responseBuffer.toString());
+          testComplete();
+        });
+      })).putHeader(HttpHeaders.ACCEPT_ENCODING, HttpHeaders.DEFLATE_GZIP).end();
+    await();
   }
 }

--- a/src/test/java/io/vertx/core/http/HttpCompressionTest.java
+++ b/src/test/java/io/vertx/core/http/HttpCompressionTest.java
@@ -55,7 +55,7 @@ public class HttpCompressionTest extends HttpTestBase {
     // server = vertx.createHttpServer();
     serverWithMinCompressionLevel = vertx.createHttpServer(serverOpts.setPort(DEFAULT_HTTP_PORT - 1).setCompressionLevel(1));
     serverWithMaxCompressionLevel = vertx.createHttpServer(serverOpts.setPort(DEFAULT_HTTP_PORT + 1).setCompressionLevel(9));
-    serverWithClientDeCompression = vertx.createHttpServer(serverOpts.setCompressionSupported(false));
+    serverWithClientDeCompression = vertx.createHttpServer(serverOpts.setCompressionSupported(true));
   }
 
   @Test
@@ -177,27 +177,8 @@ public class HttpCompressionTest extends HttpTestBase {
     String compressData = "Test Client DeCompression....";
     serverWithClientDeCompression.requestHandler(req -> {
       assertNotNull(req.headers().get("Accept-Encoding"));
-      Deflater compress = new Deflater(Deflater.DEFAULT_COMPRESSION, true);
-      byte[] bytes = compressData.getBytes();
-      compress.setInput(bytes);
-      compress.finish();
-      ByteArrayOutputStream bos = new ByteArrayOutputStream(bytes.length);
-      byte[] buf = new byte[1024];
-      byte[] output;
-      while (!compress.finished()) {
-        int i = compress.deflate(buf);
-        bos.write(buf, 0, i);
-      }
-      output = bos.toByteArray();
-      try {
-        bos.close();
-      } catch (IOException e) {
-        e.printStackTrace();
-      }
-      compress.end();
       req.response()
-        .putHeader(HttpHeaders.CONTENT_ENCODING, "deflate")
-        .end(Buffer.buffer(output));
+        .end(compressData);
     });
 
     startServer(serverWithClientDeCompression);

